File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    vector<bool> visited(computers.size(), false);
    stack<int> stk;
    
    for (int i = 0; i < computers.size(); i++) {
        
        if (visited[i]) continue;
        
        stk.push(i);
        
        while(!stk.empty()) {
            int k = stk.top();
            stk.pop();

            if (visited[k]) {
                continue;
            }

            visited[k] = true;

            for (int j = 0; j < computers[k].size(); j++) {
                if (computers[k][j] == 1 && j != k && !visited[j]) {
                    stk.push(j);
                }
            }
        }
        answer++;
        
    }    
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        int[] visited = new int[n];
            
        
        for (int i = 0; i < computers.length; i++) {
            if (visited[i] != 1) {
                dfs(computers, i, visited);
                answer++;
            }
        }
            
        return answer;
    }
    
    public void dfs(int[][] computers, int i, int[] visited) {
        visited[i] = 1;
        
        for (int j = 0; j < computers.length; j++) {
            if (i != j && computers[i][j] == 1 && visited[j] == 0) {
                dfs(computers, j, visited);
            }
        }
    }
}